### PR TITLE
Study Metadata: backend work

### PR DESCRIPTION
**Pull request by oplantalech**
This PR is the backend work for implementing the Study Metadata. This consists in giving an extra optional YAML file in the cancer study folder with extra information about the study that will show up in a popup when mousing over in a tag logo from the studies on the query page.

The YAML file with the metadata should be pointed in the `meta_study.txt` (e.g. if the name of this yaml file is `file.yml`, `meta_study.txt` needs an extra line containing `metadata_filename: file.yml`).

This PR includes:
- Changing cgds.sql and migrate.sql to add the extra column "metadata" in the table `cancer_study`
- Create endpoint to retrieve the data from the metadata column
- Change the endpoints GET /studies and GET /studies/{studyId} so that they don’t retrieve the metadata column
- Create tests for the new/changed endpoints
- Parse YAML file for metadata when importing a study
- Validate the YAML file so that it fits the format (including tests)

### CODE REVIEW

File: core/src/test/scripts/test_data/study_invalid_metadata/brca_tcga_data_cna_hg19.seg [hunk 1]
```diff
@@ -0,0 +1,10 @@
+ID	chrom	loc.start	loc.end	num.mark	seg.mean
```
Comment (pieterlukasse) [inline core/src/test/scripts/test_data/study_invalid_metadata/brca_tcga_data_cna_hg19.seg]: why this file?

File: core/src/test/scripts/test_data/study_invalid_metadata/brca_tcga_meta_cna_hg19_seg.txt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+cancer_study_identifier: study_es_0
```
Comment (pieterlukasse) [inline core/src/test/scripts/test_data/study_invalid_metadata/brca_tcga_meta_cna_hg19_seg.txt]: why this file?

File: core/src/test/scripts/test_data/study_invalid_metadata/case_lists/cases_custom.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
...	TCGA-E9-A1NC-01	TCGA-E9-A1ND-01	TCGA-E9-A1NE-01	TCGA-E9-A1NF-01	TCGA-E9-A1NG-01	TCGA-E9-A1NH-01	TCGA-E9-A1NI-01	TCGA-E9-A1QZ-01	TCGA-E9-A1R0-01	TCGA-E9-A1R2-01	TCGA-E9-A1R3-01	TCGA-E9-A1R4-01	TCGA-E9-A1R5-01	TCGA-E9-A1R6-01	TCGA-E9-A1R7-01	TCGA-E9-A1RA-01	TCGA-E9-A1RB-01	TCGA-E9-A1RC-01	TCGA-E9-A1RD-01	TCGA-E9-A1RE-01	TCGA-E9-A1RF-01	TCGA-E9-A1RG-01	TCGA-E9-A1RH-01	TCGA-E9-A1RI-01	TCGA-E9-A226-01	TCGA-E9-A227-01	TCGA-E9-A228-01	TCGA-E9-A229-01	TCGA-E9-A22A-01	TCGA-E9-A22B-01	TCGA-E9-A22D-01	TCGA-E9-A22E-01	TCGA-E9-A22G-01	TCGA-E9-A22H-01	TCGA-E9-A243-01	TCGA-E9-A244-01	TCGA-E9-A245-01	TCGA-E9-A247-01	TCGA-E9-A248-01	TCGA-E9-A249-01	TCGA-E9-A24A-01	TCGA-E9-A295-01	TCGA-EW-A1IW-01	TCGA-EW-A1IX-01	TCGA-EW-A1IY-01	TCGA-EW-A1IZ-01	TCGA-EW-A1J1-01	TCGA-EW-A1J2-01	TCGA-EW-A1J3-01	TCGA-EW-A1J5-01	TCGA-EW-A1J6-01	TCGA-EW-A1OV-01	TCGA-EW-A1OW-01	TCGA-EW-A1OX-01	TCGA-EW-A1OY-01	TCGA-EW-A1OZ-01	TCGA-EW-A1P0-01	TCGA-EW-A1P1-01	TCGA-EW-A1P3-01	TCGA-EW-A1P4-01	TCGA-EW-A1P5-01	TCGA-EW-A1P6-01	TCGA-EW-A1P7-01	TCGA-EW-A1P8-01	TCGA-EW-A1PA-01	TCGA-EW-A1PB-01	TCGA-EW-A1PD-01	TCGA-EW-A1PE-01	TCGA-EW-A1PF-01	TCGA-EW-A1PG-01	TCGA-EW-A1PH-01	TCGA-EW-A2FS-01	TCGA-EW-A2FV-01	TCGA-EW-A2FW-01	TCGA-GI-A2C8-01
```
Comment (pieterlukasse) [inline core/src/test/scripts/test_data/study_invalid_metadata/case_lists/cases_custom.txt]: why this file?

File: docs/Data-Loading.md [hunk 4]
```diff
@@ -26,6 +26,11 @@ If you want the scripts to be able to generate html reports (recommended way for
 $ sudo pip2 install jinja2
 ```
 
+To be able to use the metadata feature, you need the module `pyyaml`. You can use this command to install the module:
+```console
+$ sudo pip2 install pyyaml
```
Comment (pieterlukasse) [inline docs/Data-Loading.md]: keep in mind we are on python 3 now, so this probably needs to be updated.

File: web/pom.xml [hunk 5]
```diff
@@ -39,6 +39,12 @@
       <artifactId>json-simple</artifactId>
       <version>1.1</version>
     </dependency>
+    <!--  org.json  -->
```
Comment (pieterlukasse) [inline web/pom.xml]: I would expect we already have sufficient json libraries included / supported already. Can you please check again whether this dependency is really needed?

File: core/src/main/scripts/importer/validateData.py [hunk 6]
```diff
@@ -54,6 +54,8 @@
     # explicitly import the package, which is needed on CPython 3.4 because it
     # doesn't include https://github.com/python/cpython/pull/2639
     importlib.import_module(__package__)
+import re
+import yaml
```
Comment (sandertan) [inline core/src/main/scripts/importer/validateData.py]: `re` is already imported, please move `import yaml` up.

File: core/src/main/scripts/importer/validateData.py [hunk 7]
```diff
@@ -3963,7 +3984,11 @@ def validate_study(study_dir, portal_instance, logger, relaxed_mode, strict_maf_
                 validators_by_meta_type[
                     cbioportal_common.MetaFileTypes.PATIENT_ATTRIBUTES])})
 
-    # next validate all other data files, from meta_aaa to meta_zzz
+
```
Comment (sandertan) [inline core/src/main/scripts/importer/validateData.py]: remove a line break here.

File: core/src/main/scripts/importer/validateData.py [hunk 8]
```diff
@@ -3963,7 +3984,11 @@ def validate_study(study_dir, portal_instance, logger, relaxed_mode, strict_maf_
...
+    # then validate the metadata YAML file if it exists
+    if metadata_file_path:
+        validateMetadata(metadata_file_path, logger=logger)
+    # next validate all other data files
```
Comment (sandertan) [inline core/src/main/scripts/importer/validateData.py]: add a line break here.

File: db-scripts/src/main/resources/migration.sql [hunk 9]
```diff
@@ -454,6 +454,8 @@ ALTER TABLE gistic_to_gene DROP FOREIGN KEY gistic_to_gene_ibfk_2;
 ALTER TABLE gistic_to_gene ADD CONSTRAINT `gistic_to_gene_ibfk_2` FOREIGN KEY (`GISTIC_ROI_ID`) REFERENCES `gistic` (`GISTIC_ROI_ID`) ON DELETE CASCADE;
 UPDATE info SET DB_SCHEMA_VERSION="2.6.0";
 
+<<<<<<< HEAD
+<<<<<<< HEAD
```
Comment (sandertan) [inline db-scripts/src/main/resources/migration.sql]: Please remove this

File: db-scripts/src/main/resources/cgds.sql [hunk 10]
```diff
@@ -123,6 +123,7 @@ CREATE TABLE `cancer_study` (
   `GROUPS` varchar(200) DEFAULT NULL,
   `STATUS` int(1) DEFAULT NULL,
   `IMPORT_DATE` datetime DEFAULT NULL,
+  `METADATA` text,
```
Comment (sandertan) [inline db-scripts/src/main/resources/cgds.sql]: @pieterlukasse Should we validate the maximum size of the TEXT field type?

File: docs/Data-Loading.md [hunk 11]
```diff
@@ -27,6 +27,11 @@ If you want the scripts to be able to generate html reports (recommended way for
 $ sudo python3 -m pip install Jinja2
 ```
 
+To be able to display the [study metadata](File-Formats.md##study-metadata-file), you need the module `pyyaml`. You can use this command to install the module:
```
Comment (sandertan) [inline docs/Data-Loading.md]: Please rephrase. `pyyaml` is now needed for every validation, if it contains metadata or not, right?

File: docs/File-Formats.md [hunk 12]
```diff
@@ -1248,3 +1250,6 @@ The cells contain the p-value for the GSVA score: A real number, between 0.0 and
 <tr><td>GO_GLUCURONATE_METABOLIC_PROCES</td><td>0.6621</td><td>0.0031</td><td>1.52e-9</td></tr>
 <tr><td>..</td><td></td><td></td><td></td></tr>
 </table>
+
+## Study Metadata file
+YAML or JSON file which contains extra information (metadata) about the cancer study. No compulsory fields are required for this file (free-form). To enable this feature, you need to add a line in the cancer study meta file with `metadata_filename:` followed the YAML/JSON file name. The information on the YAML or JSON file will be displayed in a table when mousing over a tag logo in the studies on the query page.
```
Comment (sandertan) [inline docs/File-Formats.md]: JSON is also allowed? The validator only checks for correct YAML format, right? And The there is also only a unit test for YAML format. Could you add a test for JSON format then as well?

Comment (oplantalech) [reply docs/File-Formats.md]: Yes, you can use both YAML or JSON.

Comment (oplantalech) [reply docs/File-Formats.md]: > Technically YAML is a superset of JSON. This means that, in theory at least, a YAML parser can understand JSON, but not necessarily the other way around.

Which is the case here because the test I introduced checks if the metadata is not valid, and I had a hard time to find how to make it invalid 😄 But I'll introduce two tests that check that validating YAML and JSON formats for metadata do not report errors.

File: db-scripts/src/main/resources/cgds.sql (hunk 10 above)
Comment (pieterlukasse) [reply db-scripts/src/main/resources/cgds.sql]: @sandertan not really necessary, as long as the data loading step crashes with a clear error. @oplantalech can you test this with e.g. a file of 100KB?

**Comment by ersinciftci** [comment]
I think this new field can be just returned as part of the study endpoint, in `DETAILED` projection. So I don't think a new endpoint and related service/repository methods are needed.

File: core/pom.xml [hunk 13]
```diff
@@ -138,6 +138,13 @@
       <artifactId>json-simple</artifactId>
       <version>1.1</version>
     </dependency>
+    
+    <!--  org.json  -->
+    <dependency>
+        <groupId>org.json</groupId>
```
Comment (pieterlukasse) [inline core/pom.xml]: please remove this dependency, or add some comments on why it is needed. I expect we already have json libraries in this project.

File: core/src/main/java/org/mskcc/cbio/portal/util/CancerStudyReader.java [hunk 14]
```diff
@@ -97,13 +103,30 @@ private static CancerStudy getCancerStudy(TrimmedProperties properties)
...
+        
+        String metadata_filename = properties.getProperty("metadata_filename");
+        String metadata = null;
+        if (metadata_filename != null) {        	
+            try {
+                String content = new Scanner(new File(path+"/"+metadata_filename)).useDelimiter("\\Z").next();
+					Object json = new JSONParser().parse(content);
+					metadata = JSONValue.toJSONString(json);
+            } catch (FileNotFoundException e) {
+                // TODO Auto-generated catch block
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/util/CancerStudyReader.java]: throw this error

File: core/src/main/java/org/mskcc/cbio/portal/util/CancerStudyReader.java [hunk 15]
```diff
@@ -97,13 +103,30 @@ private static CancerStudy getCancerStudy(TrimmedProperties properties)
...
+                e.printStackTrace();
+            } catch (ParseException e) {
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/util/CancerStudyReader.java]: throw this error

File: core/src/main/scripts/importer/validateData.py [hunk 16]
```diff
@@ -3612,6 +3618,19 @@ def validate_defined_caselists(cancer_study_id, case_list_ids, file_types, logge
...
+        """Validate the metadata file."""
+        logger.debug('Starting validation of YAML file',
+        	extra={'filename_': metadata_file_path})
+        with open(metadata_file_path, 'r') as stream:
+            try:
+                parsedYaml = yaml.load(stream)
+                logger.info('Validation of YAML file complete.',
+                extra={'filename_': metadata_file_path})
+            except yaml.YAMLError as exc:
+                logger.error('The file provided is not in YAML format',
```
Comment (pieterlukasse) [inline core/src/main/scripts/importer/validateData.py]: if documentation states YAML/JSON, then it is perhaps also better to state it here, just for consistency.

File: docs/File-Formats.md (hunk 12 above)
Comment (pieterlukasse) [reply docs/File-Formats.md]: I think it is confusing to call this `metadata_filename` since `meta_study.txt` is already a study metadata file. Maybe change the name slightly here to `extra_metadata_filename`?

File: web/src/main/java/org/cbioportal/web/StudyController.java [hunk 17]
```diff
@@ -103,4 +111,31 @@
...
+    		} catch (JsonParseException e) {
+    			// TODO Auto-generated catch block
+    			e.printStackTrace();
+    		} catch (JsonMappingException e) {
+    			// TODO Auto-generated catch block
+    			e.printStackTrace();
+    		} catch (IOException e) {
+    			// TODO Auto-generated catch block
+    			e.printStackTrace();
+    		}
```
Comment (pieterlukasse) [inline web/src/main/java/org/cbioportal/web/StudyController.java]: all errors should be thrown

**Comment by ersinciftci** [comment]
@pieterlukasse @oplantalech I went ahead and created 2 git patch files for each option, to not cause any delay to the PR. option1 = no new endpoint, option 2 = new table. You guys can apply any option depending on the use case in your mind. If you think neither option are feasible for any reason, we can also discuss further.

[option1.txt](https://github.com/cBioPortal/cbioportal/files/2498776/option1.txt)
[option2.txt](https://github.com/cBioPortal/cbioportal/files/2498777/option2.txt)

File: core/pom.xml [hunk 18]
```diff
@@ -139,6 +139,19 @@
...
+    <!-- https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-databind -->
+    <dependency>
+        <groupId>com.fasterxml.jackson.core</groupId>
+        <artifactId>jackson-databind</artifactId>
+        <version>2.9.7</version>
+    </dependency>
+
+    <dependency>
+        <groupId>com.fasterxml.jackson.dataformat</groupId>
+        <artifactId>jackson-dataformat-yaml</artifactId>
```
Comment (pieterlukasse) [inline core/pom.xml]: we already have this in core/pom.xml:
```
    <!--yaml parser-->
    <dependency>
      <groupId>org.yaml</groupId>
      <artifactId>snakeyaml</artifactId>
      <version>1.17</version>
    </dependency>
```

Comment (pieterlukasse) [reply core/pom.xml]: please check if you can use this one, or remove it.

File: core/src/main/java/org/mskcc/cbio/portal/model/CancerStudy.java [hunk 19]
```diff
@@ -191,6 +191,7 @@ public String getCitation() {
     public void setCitation(String citation) {
         this.citation = citation;
     }
+
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/model/CancerStudy.java]: ?

File: core/src/main/java/org/mskcc/cbio/portal/util/CancerStudyReader.java [hunk 20]
```diff
@@ -40,6 +40,7 @@
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
+
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/util/CancerStudyReader.java]: ?

File: core/src/main/java/org/mskcc/cbio/portal/util/CancerStudyReader.java [hunk 21]
```diff
@@ -97,7 +98,7 @@ private static CancerStudy getCancerStudy(TrimmedProperties properties)
         if ( shortName == null) {
             throw new IllegalArgumentException("short_name is not specified.");
         }
-
+
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/util/CancerStudyReader.java]: ?

File: core/src/main/scripts/importer/validateData.py [hunk 22]
```diff
@@ -3353,7 +3354,7 @@ def process_metadata_files(directory, portal_instance, logger, relaxed_mode, str
 
     # get filenames for all meta files in the directory
     filenames = [os.path.join(directory, f) for
-                 f in os.listdir(directory) if
+                 f in sorted(os.listdir(directory)) if
```
Comment (pieterlukasse) [inline core/src/main/scripts/importer/validateData.py]: why this change?

File: db-scripts/src/main/resources/cgds.sql [hunk 23]
```diff
@@ -128,6 +128,15 @@ CREATE TABLE `cancer_study` (
   FOREIGN KEY (`TYPE_OF_CANCER_ID`) REFERENCES `type_of_cancer` (`TYPE_OF_CANCER_ID`)
 );
 
+-- --------------------------------------------------------
+CREATE TABLE `cancer_study_metadata` (
+  `CANCER_STUDY_ID` int(11) NOT NULL,
+  `CANCER_STUDY_IDENTIFIER` varchar(255),
```
Comment (pieterlukasse) [inline db-scripts/src/main/resources/cgds.sql]: why reproduce this `CANCER_STUDY_IDENTIFIER` field here? Please remove it.

File: db-scripts/src/main/resources/migration.sql [hunk 24]
```diff
@@ -538,4 +538,15 @@ AND ABS(c2.`SEGMENT_MEAN`) >= 0.2) / SUM(`END`-`START`)) AS `VALUE` FROM `copy_n
...
 HAVING SUM(`END`-`START`) > 0;
 
-UPDATE `info` SET `DB_SCHEMA_VERSION`="2.7.2";
\ No newline at end of file
+UPDATE `info` SET `DB_SCHEMA_VERSION`="2.7.2";
+
+##version: 2.7.3
+CREATE TABLE `cancer_study_metadata` (
+  `CANCER_STUDY_ID` int(11) NOT NULL,
+  `CANCER_STUDY_IDENTIFIER` varchar(255),
```
Comment (pieterlukasse) [inline db-scripts/src/main/resources/migration.sql]: same comment about `CANCER_STUDY_IDENTIFIER`

File: docs/File-Formats.md [hunk 25]
```diff
@@ -40,6 +41,7 @@ This file contains metadata about the cancer study. The file contains the follow
 7. **short_name**:  A short name used for display used on various web pages within the cBioPortal, e.g., "BRCA (Jones)".
 8. **groups (Optional)**: When using an authenticating cBioPortal, lists the user-groups that are allowed access to this study. Multiple groups are separated with a semicolon ";". The study will be invisible to users not in _at least one_ of the listed groups, as if it wasn't loaded at all. e.g., "PUBLIC;GDAC;SU2C-PI3K". see [User-Authorization](User-Authorization.md) for more information on groups
 9. **add_global_case_list (Optional)**: set to 'true' if you would like the "All samples" case list to be generated automatically for you. See also [Case lists](#case-lists).
+10. **extra_metadata_filename (Optional)**: the file name containing metadata for the [study metadata](#study-metadata-file).
```
Comment (pieterlukasse) [inline docs/File-Formats.md]: I would say:
```
the file name containing extra [study metadata](#study-metadata-file).
```

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/StudyMapper.xml [hunk 26]
```diff
@@ -21,6 +21,15 @@
             cancer_study.IMPORT_DATE AS "${prefix}importDate"
         </if>
     </sql>
+    
+    <sql id="selectMetadata">
+        cancer_study.CANCER_STUDY_ID AS "${prefix}cancerStudyId",
+        cancer_study.CANCER_STUDY_IDENTIFIER AS "${prefix}cancerStudyIdentifier"
+        <if test="projection == 'SUMMARY' || projection == 'DETAILED'">
+            ,
+            cancer_study.METADATA AS "${prefix}metadata"
```
Comment (pieterlukasse) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/StudyMapper.xml]: this will not work anymore. METADATA is not part of `cancer_study`

File: core/src/main/scripts/importer/validateData.py [hunk 27]
```diff
@@ -3370,6 +3371,7 @@ def process_metadata_files(directory, portal_instance, logger, relaxed_mode, str
     validators_by_type = {}
     case_list_suffix_fns = {}
     stable_ids = []
+    metadata_file_path = False
```
Comment (pieterlukasse) [inline core/src/main/scripts/importer/validateData.py]: is this a boolean?

File: db-scripts/src/main/resources/cgds.sql (hunk 23 above)
Comment (ersinciftci) [reply db-scripts/src/main/resources/cgds.sql]: @pieterlukasse My patch had this because I wasn't sure which way you would prefer, but it seems like you prefer joining this table with `cancer_study` in the query. (Because we get metadata by `CANCER_STUDY_IDENTIFIER` in the API). Either way works for me.

File: core/src/main/scripts/importer/validateData.py (hunk 22 above)
Comment (oplantalech) [reply core/src/main/scripts/importer/validateData.py]: To be able to pass the tests. `os.listdir` returns a list in arbitrary order (https://docs.python.org/3.5/library/os.html#os.listdir).

File: core/pom.xml (hunk 18 above)
Comment (oplantalech) [reply core/pom.xml]: I've removed this library from the pom.

**Comment by sandertan** [comment]
All my issues have been addressed.

One new issue I would like to add: The naming of this feature. cBioPortal study data contains meta, data and meta study files, and this PR introduces a "meta data" file. To us it's clear what this feature is, but to others this name will be confusing. I would consider renaming the functionality and its respective file and variable names. Perhaps "meta tag" is a better name.

**Comment by pieterlukasse** [comment]
@sandertan as discussed, perhaps "custom study tags"? This would mean the following changes need to take place:
- `cancer_study_metadata` -> `cancer_study_tags`
- `METADATA` -> `CUSTOM_TAGS`
- `extra_metadata_filename` -> `custom_tags_file`
- `/studies/{studyId}/metadata` -> `/studies/{studyId}/tags`

**Comment by oplantalech** [comment]
@ersinciftci Can you merge?

File: core/src/main/java/org/mskcc/cbio/portal/model/CancerStudy.java [hunk 28]
```diff
@@ -190,7 +190,7 @@ public String getCitation() {
 
     public void setCitation(String citation) {
         this.citation = citation;
-    }
+    }
```
[the remark this answers is not in the document]
Comment (sandertan) [reply core/src/main/java/org/mskcc/cbio/portal/model/CancerStudy.java]: @oplantalech Please remove these whites spaces.

File: core/src/test/scripts/test_data/study_es_0/meta_study.txt [hunk 29]
```diff
@@ -6,4 +6,5 @@ citation: The Hyve
 pmid: 23000897
 groups: PUBLIC;GDAC;SU2C-PI3K
 short_name: BRCA (TCGA)
-add_global_case_list: true
\ No newline at end of file
+add_global_case_list: true
+tags_file: metadata.yml
```
Comment (sandertan) [inline core/src/test/scripts/test_data/study_es_0/meta_study.txt]: Can you rename metadata.yml to `meta_tags.yml`? I considered `tags.yml` but it would be inconsistent with the other files (only data and meta files), so `meta_tags.yml` would fit right in :)

File: core/src/test/scripts/test_data/study_es_0/metadata.yml [hunk 30]
```diff
@@ -0,0 +1,18 @@
+Analyst:
```
Comment (sandertan) [inline core/src/test/scripts/test_data/study_es_0/metadata.yml]: Please rename filename, see prev comment.

File: core/src/test/scripts/test_data/study_es_0/result_report.html [hunk 31]
```diff
@@ -1912,6 +1919,52 @@ <h4 class="panel-title">meta_study.txt</h4>
...
 
+      <div class="panel panel-success">
+        <div class="panel-heading" role="tab"
+             id="heading_bWV0YWRhdGEueW1s">
+          <span class="header-icon"><span class="glyphicon glyphicon-ok" aria-hidden="true"></span><span class="sr-only">Info</span></span>
+          <a role="button" data-toggle="collapse" aria-expanded="true"
+             href="#collapse_bWV0YWRhdGEueW1s",
+             aria-controls="collapse_bWV0YWRhdGEueW1s">
+            <span class="badge">2</span>
+            <h4 class="panel-title">metadata.yml</h4>
```
Comment (sandertan) [inline core/src/test/scripts/test_data/study_es_0/result_report.html]: Please change filename.

File: core/src/test/scripts/test_data/study_es_0/result_report.html [hunk 32]
```diff
@@ -1912,6 +1919,52 @@ <h4 class="panel-title">meta_study.txt</h4>
...
+                <th>Message</th>
+                <th>Value Encountered</th>
+              </tr>
+            </thead>
+            <tbody>
+              <tr class="info">
+                <td><span class="glyphicon glyphicon-cog" aria-hidden="true"></span><span class="sr-only">Debug</span></td>
+                <td>&ndash;</td>
+                <td>&ndash;</td>
+                <td>Starting validation of YAML/JSON file</td>
```
Comment (sandertan) [inline core/src/test/scripts/test_data/study_es_0/result_report.html]: Please replace to:
Starting validation of study tags file

File: core/src/test/scripts/test_data/study_es_0/result_report.html [hunk 33]
```diff
@@ -1912,6 +1919,52 @@ <h4 class="panel-title">meta_study.txt</h4>
...
+                <td>&ndash;</td>
+              </tr>
+              <tr class="success">
+                <td><span class="glyphicon glyphicon-ok" aria-hidden="true"></span><span class="sr-only">Info</span></td>
+                <td>&ndash;</td>
+                <td>&ndash;</td>
+                <td>Validation of YAML/JSON file complete.</td>
```
Comment (sandertan) [inline core/src/test/scripts/test_data/study_es_0/result_report.html]: Replace, see above comment.

File: core/src/test/scripts/unit_tests_validate_data.py [hunk 34]
```diff
@@ -1957,7 +1957,134 @@ def test_double_cancer_type_file(self):
                          {'cancer_type_luad.txt', 'cancer_type_lung.txt'})
         # assert that the second error complains about the cancer type
         self.assertEqual(record_list[1].cause, 'luad')
-
+        
+    def test_invalid_metadata(self):
```
Comment (sandertan) [inline core/src/test/scripts/unit_tests_validate_data.py]: Please replace `metadata` with new name.

File: docs/Data-Loading.md [hunk 35]
```diff
@@ -27,6 +27,11 @@ If you want the scripts to be able to generate html reports (recommended way for
 $ sudo python3 -m pip install Jinja2
 ```
 
+The module `pyyaml` is also required (for the [study tags](File-Formats.md##study-tags-file) feature). You can use this command to install it:
+```console
+$ sudo python3 -m pip install pyyaml
+```
+
```
Comment (sandertan) [inline docs/Data-Loading.md]: Please remove this section and expend the above section:

```
$ sudo python3 -m pip install requests pyyaml
```

File: docs/File-Formats.md [hunk 36]
```diff
@@ -17,6 +17,7 @@
     * [Mutsig Data](#mutsig-data)
     * [Gene Panel Data](#gene-panel-data)
     * [Gene Set Data](#gene-set-data)
+    * [Study Metadata file](#study-metadata-file)
```
Comment (sandertan) [inline docs/File-Formats.md]: Please update this table of contents .

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java [hunk 37]
```diff
@@ -334,6 +334,25 @@ public static void addCancerStudy(CancerStudy cancerStudy, boolean overwrite) th
         reCacheAll();
     }
 
+    public static void addCancerStudyTags(CancerStudyTags cancerStudyTags) throws DaoException {
```
[the remark this answers is not in the document]
Comment (pieterlukasse) [reply core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java]: @sheridancbio I think the relevant cancer study tags should only be retrieved on demand in the frontend (different PR). E.g. when the user opens the specific view or tooltip that is meant to display the tags.

File: core/src/test/scripts/test_data/study_various_issues/result_report.html [hunk 38]
```diff
@@ -166,7 +166,7 @@ <h4 class="panel-title">General</h4>
                 <td><span class="glyphicon glyphicon-remove" aria-hidden="true"></span><span class="sr-only">Error</span></td>
                 <td>&ndash;</td>
                 <td>&ndash;</td>
-                <td>No case list found with stable_id &#39;brca_tcga_pub_all&#39;, consider adding &#39;add_global_case_list: true&#39; to the study metadata file</td>
+                <td>No case list found with stable_id &#39;brca_tcga_pub_all&#39;, consider adding &#39;add_global_case_list: true&#39; to the study tags file</td>
```
[the remark this answers is not in the document]
Comment (pieterlukasse) [reply core/src/test/scripts/test_data/study_various_issues/result_report.html]: yes, this is wrong. This means the warning given by the validation is now wrong. @oplantalech please open a new PR to fix this.

File: db-scripts/src/main/resources/cgds.sql [hunk 39]
```diff
@@ -128,6 +128,14 @@ CREATE TABLE `cancer_study` (
   FOREIGN KEY (`TYPE_OF_CANCER_ID`) REFERENCES `type_of_cancer` (`TYPE_OF_CANCER_ID`)
 );
 
+-- --------------------------------------------------------
+CREATE TABLE `cancer_study_tags` (
```
[the remark this answers is not in the document]
Comment (pieterlukasse) [reply db-scripts/src/main/resources/cgds.sql]: good catch. @oplantalech please fix this one as well.

File: db-scripts/src/main/resources/cgds.sql [hunk 40]
```diff
@@ -128,6 +128,14 @@ CREATE TABLE `cancer_study` (
...
+  `CANCER_STUDY_ID` int(11) NOT NULL,
+  `TAGS` text,
+  UNIQUE (`CANCER_STUDY_ID`),
```
[the remark this answers is not in the document]
Comment (pieterlukasse) [reply db-scripts/src/main/resources/cgds.sql]: @oplantalech can you check this one?

File: model/src/main/java/org/cbioportal/model/CancerStudyTags.java [hunk 41]
```diff
@@ -0,0 +1,25 @@
...
+
+    public Integer getCancerStudyId() {
+        return cancerStudyId;
+    }
+
+    public void setCancerStudyId(Integer cancerStudyId) {
+        this.cancerStudyId = cancerStudyId;
+    }
+
+    public String getTags() {
```
[the remark this answers is not in the document]
Comment (pieterlukasse) [reply model/src/main/java/org/cbioportal/model/CancerStudyTags.java]: it is validated upon import.

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/StudyMapper.xml [hunk 42]
```diff
@@ -110,5 +110,15 @@
...
         GROUP BY cancer_study.CANCER_STUDY_ID
     </select>
+    
+    <select id="getTags" resultType="org.cbioportal.model.CancerStudyTags">
+        SELECT
+        cancer_study_tags.CANCER_STUDY_ID AS cancerStudyId,
+        cancer_study_tags.TAGS AS tags
+        FROM cancer_study_tags
+        WHERE cancer_study_tags.CANCER_STUDY_ID in (
+            SELECT cancer_study.CANCER_STUDY_ID FROM cancer_study WHERE cancer_study.CANCER_STUDY_IDENTIFIER = #{studyId}
```
[the remark this answers is not in the document]
Comment (pieterlukasse) [reply persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/StudyMapper.xml]: `CANCER_STUDY_IDENTIFIER` is unique (see https://github.com/cBioPortal/cbioportal/blob/master/db-scripts/src/main/resources/cgds.sql#L127). So this query is confusing. A simple join would be clearer. @oplantalech please take this one as well.
